src/Config/Database/DbConfig.php

```diff
@@ -0,0 +1,298 @@
+     *
+     * @var RelationshipConnectionFactory
+     */
+    private $connectionDataFactory;
+
+    /**
+     * Final database configuration after merging
+     *
+     * @var array
+     */
+    private $dbConfig;
+
+    /**
+     * @param DeployInterface $stageConfig
+     * @param ConfigMerger $configMerger
+     * @param RelationshipConnectionFactory $connectionDataFactory
+     */
+    public function __construct(
+        DeployInterface $stageConfig,
+        ConfigMerger $configMerger,
+        RelationshipConnectionFactory $connectionDataFactory
+    ) {
+        $this->stageConfig = $stageConfig;
+        $this->configMerger = $configMerger;
+        $this->connectionDataFactory = $connectionDataFactory;
+    }
+
+    /**
+     * Returns database configurations as merge result of customer configurations from .magento.env.yaml
+     * and connection data from cloud environment.
+     * Result is in the format which is used in magento env.php file.
+     * Result may contain 2 elements: 'connection' and/or 'slave_connection'
+     *
+     * @return array
+     * ```php
+     * [
+     *   'connection' => [
+     *     'default' => [
+     *       'host' => {host},
+     *       ...
+     *     ],
+     *   ],
+     *   'slave_connection' => [...]
+     * ]
+     * ```
+     */
+    public function get(): array
+    {
+        if (null !== $this->dbConfig) {
+            return $this->dbConfig;
+        }
+
+        $customConfig = $this->getSupportedCustomConfig();
+
+        if (!$this->configMerger->isEmpty($customConfig) && !$this->configMerger->isMergeRequired($customConfig)) {
+            return $this->configMerger->clear($customConfig);
+        }
+
+        return $this->dbConfig = $this->configMerger->merge($this->collectEnvConfig($customConfig), $customConfig);
+    }
+
+    /**
+     *  Collects DB connections which exist in environment.
+     *
+     *
+     * @param array $customDbConfig
+     * @return array
+     */
+    private function collectEnvConfig(array $customDbConfig): array
+    {
+        $config = [];
+        foreach (self::MAIN_CONNECTION_MAP as $mageConnName => $envConnName) {
+            // collect main connections
+            $envDbConfig = $this->getEnvConnectionData($envConnName);
+            if (empty($envDbConfig->getHost())) {
+                continue;
+            }
+            $config[self::KEY_CONNECTION][$mageConnName] = $this->convertToMageFormat(
+                $envDbConfig,
+                !in_array($mageConnName, self::MAIN_CONNECTIONS)
+            );
+            // collect slave connections
+            if (!isset(self::SLAVE_CONNECTION_MAP[$mageConnName])) {
+                continue;
+            }
+            $slaveConnectionData = $this->getEnvConnectionData(self::SLAVE_CONNECTION_MAP[$mageConnName]);
+            if (empty($slaveConnectionData->getHost())
+                || !$this->isCustomConnectionCompatibleForSlave($customDbConfig, $mageConnName, $envDbConfig)) {
```

Comment:
It would be nice to write some messages into the log in the case when config is not compatible.

Reply:
It's here https://github.com/magento/ece-tools/pull/641/files#diff-e2fd2316780e64f5f9ea05a89e9925a8R302